share/system/cgroup_linux.go

```diff
@@ -332,76 +332,147 @@ func getCgroupPath_cgroup_v2(pid int) (string, error) {
+	scanner := bufio.NewScanner(f)
+	for scanner.Scan() {
+
+		tokens := strings.Split(scanner.Text(), ":")
+
+		if len(tokens) > 2 {
+			// https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/7/html/resource_management_guide/sec-default_cgroup_hierarchies
+			// https://docs.fedoraproject.org/en-US/quick-docs/understanding-and-administering-systemd/
+			// cgroup path is defined by <index>:<subsystem(s)>:<path>
+			// example: 2:cpuset,cpu,cpuacct,memory,net_cls,net_prio,hugetlb:/kubepods/besteffort/pod1fe19bf5-e8ef-11e8-900c-52daee5a874d/da31e536c8d61304a6d5998d163d12400a7a9a1003e1d86369e8fadb022fc17d
+
+			// The subsystem is the second item in the line
+			path := tokens[2]
+
+			// handle multiple subsytem keys that use the same path
+			// eg:  7:cpu,cpuacct:/docker/b5827d5acf95f5b286ae4aa28718162a3ed2152e7f4d4048dc9d2456540c11ce
+			subs := strings.Split(tokens[1], ",")
+
+			for _, sub := range subs {
+				subsystemMap[sub] = path
+			}
 		}
-		return path, nil
+	}
 
-	case cgroup_v2:
-		// unified file structure
-		return getCgroupPath_cgroup_v2(pid)
+	// On docker k8s (or a host cgroup file), it only returns a single line so we will return the first item
+	if len(subsystemMap) == 1 {
+		for _, val := range subsystemMap {
+			return CgroupRelativePathFix(val), nil
+		}
+	}
+
+	// Match the subsystem we're interested in.
+	if cpath, ok := subsystemMap[subsystem]; ok {
+		return CgroupRelativePathFix(cpath), nil
+	}
+
+	// If the subsystem is not available, then we will have to fallback
+	return defaultHostCgroup, fmt.Errorf("[%s] subsystem not found in tokens: %+v", subsystem, subsystemMap)
+}
+
+// CgroupRelativePathFix - Applies fixes for cgroup files with certain types of paths
+func CgroupRelativePathFix(path string) (string) {
+	if strings.Contains(path, "/..") {
+		// I can't find the documentation why the cgroup file will hold relative paths. For now, I'm applying these
+		// fixes as we encounter them. Documentation on this behaviour would make this code more robust.
+		// Example: 0::/../../kubepods-besteffort-poddee9029c_408f_4466_811d_43eea3042395.slice/docker-a737350ff4843bb79debc4e2dc98f0b1b11d40f814ea4303d9167dd70c314b95.scope
+
+		path  = filepath.Clean(path)
+		if strings.Contains(path, "kubepods-pod") {
+			path = filepath.Join("/kubepods.slice", path)
+		} else if strings.Contains(path, "kubepods-besteffort") {
+			path = filepath.Join("/kubepods.slice/kubepods-besteffort.slice", path)
+		} else if strings.Contains(path, "kubepods-burstable") {
+			path = filepath.Join("/kubepods.slice", path)
+		}
+	}
+	return path
+}
+
+// cgroup v2 is collected inside an unified file folder
+func (s *SystemTools) getCgroupMetricsPath(pid int, subsystem string) (string, error) {
+	var path string
+
+	if pid == 0 { // self
+		path = "/proc/self/cgroup"
+	} else {
+		path = filepath.Join(s.procDir, strconv.Itoa(pid), "cgroup")
 	}
 
-	// However, the k8s POD does not have those subsystem folders
```

Comment:
In the original code, this logic would never get executed, we would always be in v1 or v2 mode.